solidity/contracts/PhasedEscrow.sol

```diff
@@ -70,26 +70,127 @@ contract PhasedEscrow is Ownable {
     }
 }
 
-interface ICurveRewards {
-    function notifyRewardAmount(uint256 amount) external;
+/// @title BatchedPhasedEscrow
+/// @notice A token holder contract allowing contract owner to approve a set of
+///         beneficiaries of tokens held by the contract, to appoint a separate
+///         drawee role, and allowing that drawee to withdraw tokens to approved
+///         beneficiaries in phases.
+contract BatchedPhasedEscrow is Ownable {
+    using SafeERC20 for IERC20;
+
+    event BeneficiaryApproved(address beneficiary);
+    event TokensWithdrawn(address beneficiary, uint256 amount);
+    event DraweeRoleTransferred(address oldDrawee, address newDrawee);
+
+    IERC20 public token;
+    address public drawee;
+    mapping(address => bool) private approvedBeneficiaries;
+
+    modifier onlyDrawee() {
+        require(drawee == msg.sender, "Caller is not the drawee");
+        _;
+    }
+
+    constructor(IERC20 _token) public {
+        token = _token;
+        drawee = msg.sender;
+    }
+
+    /// @notice Funds the escrow by transferring all of the approved tokens
+    ///         to the escrow.
+    function receiveApproval(
+        address _from,
+        uint256 _value,
+        address _token,
+        bytes memory
+    ) public {
+        require(IERC20(_token) == token, "Unsupported token");
+        token.safeTransferFrom(_from, address(this), _value);
+    }
+
+    /// @notice Approves the provided address as a beneficiary of tokens held by
+    ///         the escrow. Can be called only by escrow owner.
+    function approveBeneficiary(
+        IBeneficiaryContract _beneficiary
+    ) external onlyOwner {
+        address beneficiaryAddress = address(_beneficiary);
+        require(
+            beneficiaryAddress != address(0), 
+            "Beneficiary can not be zero address"
+        );
+        approvedBeneficiaries[beneficiaryAddress] = true;
+        emit BeneficiaryApproved(beneficiaryAddress);
+    }
+
+    /// @notice Returns `true` if the given address has been approved as a
+    ///         beneficiary of the escrow, `false` otherwise.
+    function isBeneficiaryApproved(
+        IBeneficiaryContract _beneficiary
+    ) external view returns (bool) {
+        return approvedBeneficiaries[address(_beneficiary)];
+    }
+
+    /// @notice Transfers the role of drawee to another address. Can be called
+    ///         only by the contract owner.
+    function transferDraweeRole(address newDrawee) public onlyOwner {
```

Comment:
Thoughts on renaming `transferDraweeRole` to `setDraweeRole`? Having in mind how `transferOwnership` from `Ownable` works, the current name suggest that this function transfers the role from the `msg.sender` to the `newDrawee`.

Reply:
Hmm... Yes, very good point. Will update.

Reply:
Actually, I think it should be just `setDrawee` - we are setting the property, that's all. We were transferring the role but now we just *set* who is a drawee. WDYT?

Reply:
YES! 💪🏻

Reply:
Perfect. Done in https://github.com/keep-network/keep-core/pull/2230/commits/e63422335bbbac248d1ba6add95c473d540e6a36.